File: frontend-modern/src/api/patrol.ts
/**
 * Pulse Patrol API client
 * Provides access to background AI monitoring findings and status
 */

import { apiFetchJSON } from '@/utils/apiClient';

export type FindingSeverity = 'info' | 'watch' | 'warning' | 'critical';
export type FindingCategory = 'performance' | 'capacity' | 'reliability' | 'backup' | 'security' | 'general';

export interface Finding {
    id: string;
    key?: string;
    severity: FindingSeverity;
    category: FindingCategory;
    resource_id: string;
    resource_name: string;
    resource_type: string; // node, vm, container, docker_host, docker_container, storage, pbs, host_raid
    node?: string;
    title: string;
    description: string;
    recommendation?: string;
    evidence?: string;
    detected_at: string;
    last_seen_at: string;
    resolved_at?: string;
    auto_resolved: boolean;
    acknowledged_at?: string;
    snoozed_until?: string; // Finding hidden until this time
    alert_id?: string;
    // User feedback fields (LLM memory system)
    dismissed_reason?: 'not_an_issue' | 'expected_behavior' | 'will_fix_later';
    user_note?: string;
    times_raised: number;
    suppressed: boolean;
    // Investigation fields (Patrol Autonomy)
    investigation_session_id?: string;
    investigation_status?: InvestigationStatus;
    investigation_outcome?: InvestigationOutcome;
    last_investigated_at?: string;
    investigation_attempts: number;
}

export type InvestigationStatus = 'pending' | 'running' | 'completed' | 'failed' | 'needs_attention';
export type InvestigationOutcome =
    | 'resolved'
    | 'fix_queued'
    | 'fix_executed'
    | 'fix_failed'
    | 'needs_attention'
    | 'cannot_fix'
    | 'timed_out'
    | 'fix_verified'
    | 'fix_verification_failed'
    | 'fix_verification_unknown';
export type PatrolAutonomyLevel = 'monitor' | 'approval' | 'assisted' | 'full';

export interface PatrolAutonomySettings {
    autonomy_level: PatrolAutonomyLevel;
    full_mode_unlocked: boolean;       // User has acknowledged Full mode risks
    investigation_budget: number;      // Max turns per investigation (5-30)
    investigation_timeout_sec: number; // Max seconds per investigation (60-600)
}

export interface Investigation {
    id: string;
    finding_id: string;
    session_id: string;
    status: InvestigationStatus;
    started_at: string;
    completed_at?: string;
    turn_count: number;
    outcome?: InvestigationOutcome;
    tools_available?: string[];
    tools_used?: string[];
    evidence_ids?: string[];
    summary?: string;
    error?: string;
    proposed_fix?: ProposedFix;
    approval_id?: string;
}

export interface ProposedFix {
    id: string;
    description: string;
    commands?: string[];
    risk_level?: string;
    destructive: boolean;
    target_host?: string;
    rationale?: string;
}

export interface InvestigationMessages {
    investigation_id: string;
    session_id: string;
    messages: ChatMessage[];
}

export interface ChatMessage {
    id: string;
    role: 'user' | 'assistant' | 'system';
    content: string;
    reasoning_content?: string;
    tool_calls?: ChatToolCall[];
    tool_result?: ChatToolResult;
    timestamp: string;
}

export interface ChatToolCall {
    id: string;
    name: string;
    input: Record<string, unknown>;
}

export interface ChatToolResult {
    tool_use_id: string;
    content: string;
    is_error?: boolean;
}

export interface FindingsSummary {
    critical: number;
    warning: number;
    watch: number;
    info: number;
}

export type LicenseStatus = 'none' | 'active' | 'expired' | 'grace_period';

export interface PatrolStatus {
    running: boolean;
    enabled: boolean;
    last_patrol_at?: string;
    next_patrol_at?: string;
    last_duration_ms: number;
    resources_checked: number;
    findings_count: number;
    error_count: number;
    healthy: boolean;
    interval_ms: number; // Patrol interval in milliseconds
    fixed_count: number; // Number of issues auto-fixed by Patrol
    blocked_reason?: string;
    blocked_at?: string;
    license_required?: boolean;
    license_status?: LicenseStatus;
    upgrade_url?: string;
    summary: FindingsSummary;
}

/**
 * Get the current Pulse Patrol status
 */
export async function getPatrolStatus(): Promise<PatrolStatus> {
    return apiFetchJSON<PatrolStatus>('/api/ai/patrol/status');
}

/**
 * Acknowledge a finding (marks as seen but keeps visible, like alert acknowledgement)
 * Finding will auto-resolve when the underlying condition clears.
 */
export async function acknowledgeFinding(findingId: string): Promise<{ success: boolean; message: string }> {
    return apiFetchJSON('/api/ai/patrol/acknowledge', {
        method: 'POST',
        body: JSON.stringify({ finding_id: findingId }),
    });
}

/**
 * Snooze a finding for a specified duration
 * @param findingId The ID of the finding to snooze
 * @param durationHours Duration in hours (e.g., 1, 24, 168 for 7 days)
 */
export async function snoozeFinding(findingId: string, durationHours: number): Promise<{ success: boolean; message: string }> {
    return apiFetchJSON('/api/ai/patrol/snooze', {
        method: 'POST',
        body: JSON.stringify({ finding_id: findingId, duration_hours: durationHours }),
    });
}

/**
 * Dismiss a finding with a reason (LLM memory feature)
 * The LLM will be told not to re-raise this issue in future patrols.
 * @param findingId The ID of the finding to dismiss
 * @param reason One of: "not_an_issue", "expected_behavior", "will_fix_later"
 * @param note Optional freeform explanation
 */
export async function dismissFinding(
    findingId: string,
    reason: 'not_an_issue' | 'expected_behavior' | 'will_fix_later',
    note?: string
): Promise<{ success: boolean; message: string }> {
    return apiFetchJSON('/api/ai/patrol/dismiss', {
        method: 'POST',
        body: JSON.stringify({ finding_id: findingId, reason, note }),
    });
}

/**
 * Set or update a user note on a finding
 * Notes provide context that Patrol sees on future runs.
 * @param findingId The ID of the finding
 * @param note The note text (empty string to clear)
 */
export async function setFindingNote(findingId: string, note: string): Promise<{ success: boolean; message: string }> {
    return apiFetchJSON('/api/ai/patrol/findings/note', {
        method: 'POST',
        body: JSON.stringify({ finding_id: findingId, note }),
    });
}

// =============================================================================
// Patrol Autonomy APIs
// =============================================================================

/**
 * Get current patrol autonomy settings
 */
export async function getPatrolAutonomySettings(): Promise<PatrolAutonomySettings> {
    return apiFetchJSON<PatrolAutonomySettings>('/api/ai/patrol/autonomy');
}

/**
 * Update patrol autonomy settings
 */
export async function updatePatrolAutonomySettings(settings: PatrolAutonomySettings): Promise<{ success: boolean; settings: PatrolAutonomySettings }> {
    return apiFetchJSON('/api/ai/patrol/autonomy', {
        method: 'PUT',
        body: JSON.stringify(settings),
    });
}

/**
 * Get investigation details for a finding
 */
export async function getInvestigation(findingId: string): Promise<Investigation> {
    return apiFetchJSON<Investigation>(`/api/ai/findings/${encodeURIComponent(findingId)}/investigation`);
}

/**
 * Get chat messages for an investigation
 */
export async function getInvestigationMessages(findingId: string): Promise<InvestigationMessages> {
    return apiFetchJSON<InvestigationMessages>(`/api/ai/findings/${encodeURIComponent(findingId)}/investigation/messages`);
}

/**
 * Trigger re-investigation of a finding
 */
export async function reinvestigateFinding(findingId: string): Promise<{ success: boolean; message: string }> {
    return apiFetchJSON(`/api/ai/findings/${encodeURIComponent(findingId)}/reinvestigate`, {
        method: 'POST',
    });
}

/**
 * Severity color mapping for UI
 */
export const severityColors: Record<FindingSeverity, { bg: string; text: string; border: string }> = {
    critical: { bg: 'rgba(220, 38, 38, 0.15)', text: '#ef4444', border: 'rgba(220, 38, 38, 0.3)' },
    warning: { bg: 'rgba(234, 179, 8, 0.15)', text: '#eab308', border: 'rgba(234, 179, 8, 0.3)' },
    watch: { bg: 'rgba(59, 130, 246, 0.15)', text: '#3b82f6', border: 'rgba(59, 130, 246, 0.3)' },
    info: { bg: 'rgba(107, 114, 128, 0.15)', text: '#9ca3af', border: 'rgba(107, 114, 128, 0.3)' },
};

/**
 * Format a timestamp for display
 */
export function formatTimestamp(ts: string): string {
    const date = new Date(ts);
    if (!ts || Number.isNaN(date.getTime())) {
        return '';
    }
    const now = new Date();
    const diffMs = now.getTime() - date.getTime();
    const diffMins = Math.floor(diffMs / 60000);
    const diffHours = Math.floor(diffMs / 3600000);
    const diffDays = Math.floor(diffMs / 86400000);

    if (diffMins < 1) return 'just now';
    if (diffMins < 60) return `${diffMins}m ago`;
    if (diffHours < 24) return `${diffHours}h ago`;
    if (diffDays < 7) return `${diffDays}d ago`;
    return date.toLocaleDateString();
}

/**
 * Investigation status labels for UI
 */
export const investigationStatusLabels: Record<InvestigationStatus, string> = {
    pending: 'Pending',
    running: 'Investigating...',
    completed: 'Completed',
    failed: 'Failed',
    needs_attention: 'Needs Attention',
};

/**
 * Investigation outcome labels for UI
 */
export const investigationOutcomeLabels: Record<InvestigationOutcome, string> = {
    resolved: 'Resolved',
    fix_queued: 'Fix Queued',
    fix_executed: 'Fix Executed',
    fix_failed: 'Fix Failed',
    needs_attention: 'Needs Attention',
    cannot_fix: 'Cannot Auto-Fix',
    timed_out: 'Timed Out — Will Retry',
    fix_verified: 'Fix Verified',
    fix_verification_failed: 'Verification Failed',
    fix_verification_unknown: 'Verification Inconclusive',
};

/**
 * Investigation outcome badge colors for UI
 */
export const investigationOutcomeColors: Record<InvestigationOutcome, string> = {
    resolved: 'border-green-200 bg-green-50 text-green-700 dark:border-green-800 dark:bg-green-900 dark:text-green-300',
    fix_queued: 'border-blue-200 bg-blue-50 text-blue-700 dark:border-blue-800 dark:bg-blue-900 dark:text-blue-300',
    fix_executed: 'border-green-200 bg-green-50 text-green-700 dark:border-green-800 dark:bg-green-900 dark:text-green-300',
    fix_failed: 'border-red-200 bg-red-50 text-red-700 dark:border-red-800 dark:bg-red-900 dark:text-red-300',
    needs_attention: 'border-amber-200 bg-amber-50 text-amber-700 dark:border-amber-800 dark:bg-amber-900 dark:text-amber-300',
    cannot_fix: 'border-slate-200 bg-slate-50 text-slate-600',
    timed_out: 'border-amber-200 bg-amber-50 text-amber-700 dark:border-amber-800 dark:bg-amber-900 dark:text-amber-300',
    fix_verified: 'border-green-200 bg-green-50 text-green-700 dark:border-green-800 dark:bg-green-900 dark:text-green-300',
    fix_verification_failed: 'border-red-200 bg-red-50 text-red-700 dark:border-red-800 dark:bg-red-900 dark:text-red-300',
    fix_verification_unknown: 'border-amber-200 bg-amber-50 text-amber-700 dark:border-amber-800 dark:bg-amber-900 dark:text-amber-300',
};

// =============================================================================
// Patrol Run History APIs
// =============================================================================

export type PatrolRunStatus = 'healthy' | 'issues_found' | 'critical' | 'error';

export interface ToolCallRecord {
    id: string;
    tool_name: string;
    input: string;
    output: string;
    success: boolean;
    start_time: number;
    end_time: number;
    duration_ms: number;
}

export interface PatrolRunRecord {
    id: string;
    started_at: string;
    completed_at: string;
    duration_ms: number;
    type: string;
    trigger_reason?: string;
    scope_resource_ids?: string[];
    scope_resource_types?: string[];
    scope_context?: string;
    alert_id?: string;
    finding_id?: string;
    resources_checked: number;
    nodes_checked: number;
    guests_checked: number;
    docker_checked: number;
    storage_checked: number;
    hosts_checked: number;
    pbs_checked: number;
    pmg_checked: number;
    kubernetes_checked: number;
    new_findings: number;
    existing_findings: number;
    rejected_findings: number;
    resolved_findings: number;
    auto_fix_count: number;
    findings_summary: string;
    finding_ids: string[];
    error_count: number;
    status: PatrolRunStatus;
    ai_analysis?: string;
    input_tokens?: number;
    output_tokens?: number;
    tool_calls?: ToolCallRecord[];
    tool_call_count: number;
}

function normalizeHistoryLimit(limit: number): number {
    if (!Number.isFinite(limit)) {
        return 30;
    }
    const normalized = Math.floor(limit);
    if (normalized < 1) {
        return 1;
    }
    return normalized;
}

/**
 * Get patrol run history
 * @param limit Maximum number of runs to return (default: 30)
 */
export async function getPatrolRunHistory(limit: number = 30): Promise<PatrolRunRecord[]> {
    const search = new URLSearchParams({
        limit: String(normalizeHistoryLimit(limit)),
    });
    const runs = await apiFetchJSON<PatrolRunRecord[]>(`/api/ai/patrol/runs?${search.toString()}`);
    return runs || [];
}

/**
 * Get patrol run history with tool calls included
 * @param limit Maximum number of runs to return (default: 30)
 */
export async function getPatrolRunHistoryWithToolCalls(limit: number = 30): Promise<PatrolRunRecord[]> {
    const search = new URLSearchParams({
        include: 'tool_calls',
        limit: String(normalizeHistoryLimit(limit)),
    });
    const runs = await apiFetchJSON<PatrolRunRecord[]>(`/api/ai/patrol/runs?${search.toString()}`);
    return runs || [];
}

/** SSE event from /api/ai/patrol/stream */
export interface PatrolStreamEvent {
    // Meta (best-effort; present when backend supports it)
    run_id?: string;
    seq?: number;
    ts_ms?: number;
    resync_reason?: 'late_joiner' | 'stale_last_event_id' | 'buffer_rotated';
    buffer_start_seq?: number;
    buffer_end_seq?: number;
    content_truncated?: boolean;

    type: 'snapshot' | 'start' | 'content' | 'phase' | 'thinking' | 'complete' | 'error' | 'tool_start' | 'tool_end';
    content?: string;
    phase?: string;
    tokens?: number;
    tool_id?: string;
    tool_name?: string;
    tool_input?: string;
    tool_raw_input?: string;
    tool_output?: string;
    tool_success?: boolean;
}

/**
 * Trigger a manual patrol run
 */
export async function triggerPatrolRun(): Promise<{ success: boolean; message: string }> {
    return apiFetchJSON('/api/ai/patrol/run', {
        method: 'POST',
    });
}

File: internal/ai/patrol.go
// patrol.go defines the core types, interfaces, and struct for the PatrolService.
//
// Architecture:
//
//	Scheduled/Event Trigger
//	        │
//	        ▼
//	buildSeedContext()  ── infrastructure snapshot
//	        │
//	        ▼
//	runAIAnalysis()     ── agentic LLM loop with tools
//	        │
//	        ▼
//	recordFinding()     ── dedup, threshold validation
//	        │
//	        ├──▶ MaybeInvestigateFinding()  ── autonomous investigation
//	        │         │
//	        │         ▼
//	        │    parseInvestigationSummary() ── extract PROPOSED_FIX
//	        │         │
//	        │         ▼
//	        │    approval / execution / verification
//	        │
//	        └──▶ generateRemediationPlan()  ── template-based fix plan
//
// Safety: All command execution goes through internal/ai/safety for
// blocked command detection. Investigation guardrails and remediation
// engine both delegate to the shared safety package.
//
// The patrol system is split across these files:
//   - patrol.go: types, interfaces, PatrolService struct, constructor
//   - patrol_init.go: configuration, setters, getters, dependency injection
//   - patrol_run.go: lifecycle, scheduling, streaming, alert resolution
//   - patrol_ai.go: LLM interaction, seed context, prompt construction
//   - patrol_findings.go: finding lifecycle, remediation planning, investigation
//   - patrol_signals.go: deterministic signal detection from tool call outputs
//   - patrol_metrics.go: metrics collection for patrol operations
package ai

import (
	"context"
	"sync"
	"sync/atomic"
	"time"

	"github.com/rcourtman/pulse-go-rewrite/internal/ai/baseline"
	"github.com/rcourtman/pulse-go-rewrite/internal/ai/circuit"
	"github.com/rcourtman/pulse-go-rewrite/internal/ai/finding"
	"github.com/rcourtman/pulse-go-rewrite/internal/ai/knowledge"
	"github.com/rcourtman/pulse-go-rewrite/internal/ai/memory"
	"github.com/rcourtman/pulse-go-rewrite/internal/ai/remediation"
	"github.com/rcourtman/pulse-go-rewrite/internal/alerts"
	"github.com/rcourtman/pulse-go-rewrite/internal/relay"
	"github.com/rcourtman/pulse-go-rewrite/internal/servicediscovery"
	"github.com/rcourtman/pulse-go-rewrite/internal/unifiedresources"
)

// ThresholdProvider provides user-configured alert thresholds for patrol to use
type ThresholdProvider interface {
	// GetNodeCPUThreshold returns the CPU alert trigger threshold for nodes (0-100%)
	GetNodeCPUThreshold() float64
	// GetNodeMemoryThreshold returns the memory alert trigger threshold for nodes (0-100%)
	GetNodeMemoryThreshold() float64
	// GetGuestMemoryThreshold returns the memory alert trigger threshold for guests (0-100%)
	GetGuestMemoryThreshold() float64
	// GetGuestDiskThreshold returns the disk alert trigger threshold for guests (0-100%)
	GetGuestDiskThreshold() float64
	// GetStorageThreshold returns the usage alert trigger threshold for storage (0-100%)
	GetStorageThreshold() float64
}

// AlertResolver provides the ability to review and resolve alerts
type AlertResolver interface {
	// GetActiveAlerts returns all currently active alerts
	GetActiveAlerts() []AlertInfo
	// ResolveAlert clears an active alert, returns true if successful
	ResolveAlert(alertID string) bool
}

// PatrolStatus represents the current state of the patrol service
type PatrolStatus struct {
	Running          bool          `json:"running"`
	Enabled          bool          `json:"enabled"`
	LastPatrolAt     *time.Time    `json:"last_patrol_at,omitempty"`
	NextPatrolAt     *time.Time    `json:"next_patrol_at,omitempty"`
	LastDuration     time.Duration `json:"last_duration_ms"`
	ResourcesChecked int           `json:"resources_checked"`
	FindingsCount    int           `json:"findings_count"`
	ErrorCount       int           `json:"error_count"`
	Healthy          bool          `json:"healthy"`
	IntervalMs       int64         `json:"interval_ms"` // Patrol interval in milliseconds
	BlockedReason    string        `json:"blocked_reason,omitempty"`
	BlockedAt        *time.Time    `json:"blocked_at,omitempty"`
}

// PatrolRunRecord represents a single patrol check run
type PatrolRunRecord struct {
	ID                 string        `json:"id"`
	StartedAt          time.Time     `json:"started_at"`
	CompletedAt        time.Time     `json:"completed_at"`
	Duration           time.Duration `json:"-"`
	DurationMs         int64         `json:"duration_ms"`
	Type               string        `json:"type"` // Always "patrol" now (kept for backwards compat)
	TriggerReason      string        `json:"trigger_reason,omitempty"`
	ScopeResourceIDs   []string      `json:"scope_resource_ids,omitempty"`
	ScopeResourceTypes []string      `json:"scope_resource_types,omitempty"`
	ScopeContext       string        `json:"scope_context,omitempty"`
	AlertID            string        `json:"alert_id,omitempty"`
	FindingID          string        `json:"finding_id,omitempty"`
	ResourcesChecked   int           `json:"resources_checked"`
	// Breakdown by resource type
	NodesChecked      int `json:"nodes_checked"`
	GuestsChecked     int `json:"guests_checked"`
	DockerChecked     int `json:"docker_checked"`
	StorageChecked    int `json:"storage_checked"`
	HostsChecked      int `json:"hosts_checked"`
	PBSChecked        int `json:"pbs_checked"`
	PMGChecked        int `json:"pmg_checked"`
	KubernetesChecked int `json:"kubernetes_checked"`
	// Findings from this run
	NewFindings      int      `json:"new_findings"`
	ExistingFindings int      `json:"existing_findings"`
	RejectedFindings int      `json:"rejected_findings"`
	ResolvedFindings int      `json:"resolved_findings"`
	AutoFixCount     int      `json:"auto_fix_count,omitempty"`
	FindingsSummary  string   `json:"findings_summary"` // e.g., "All healthy" or "2 warnings, 1 critical"
	FindingIDs       []string `json:"finding_ids"`      // IDs of findings from this run
	ErrorCount       int      `json:"error_count"`
	Status           string   `json:"status"` // "healthy", "issues_found", "error"
	// AI Analysis details
	AIAnalysis   string `json:"ai_analysis,omitempty"`   // The AI's raw response/analysis
	InputTokens  int    `json:"input_tokens,omitempty"`  // Tokens sent to AI
	OutputTokens int    `json:"output_tokens,omitempty"` // Tokens received from AI
	// Tool call traces
	ToolCalls     []ToolCallRecord `json:"tool_calls,omitempty"`
	ToolCallCount int              `json:"tool_call_count"`
}

// MaxPatrolRunHistory is the maximum number of patrol runs to keep in history
const MaxPatrolRunHistory = 100

const (
	MaxToolInputSize   = 1024 // max chars for tool input in persisted record
	MaxToolOutputSize  = 2048 // max chars for tool output in persisted record
	MaxToolCallsPerRun = 100  // max tool calls stored per run (oldest dropped)
)

// LearningProvider provides learned preferences for patrol context
type LearningProvider interface {
	// FormatForContext returns learned preferences formatted for AI prompt injection
	FormatForContext() string
}

// ProxmoxEventProvider provides recent Proxmox events for patrol context
type ProxmoxEventProvider interface {
	// FormatForPatrol formats recent events for AI patrol context
	FormatForPatrol(duration time.Duration) string
}

// ForecastProvider provides trend forecasts for patrol context
type ForecastProvider interface {
	// FormatKeyForecasts returns formatted forecasts for resources with concerning trends
	FormatKeyForecasts() string
}

// UnifiedFindingCallback is called when patrol creates a new finding
// It allows the unified store to receive patrol findings in addition to alerts
type UnifiedFindingCallback func(f *Finding) bool

// PushNotifyCallback is called to send a push notification through the relay.
type PushNotifyCallback func(notification relay.PushNotificationPayload)

// InvestigationOrchestrator defines the interface for autonomous investigation of findings
type InvestigationOrchestrator interface {
	// InvestigateFinding starts an investigation for a finding
	InvestigateFinding(ctx context.Context, finding *InvestigationFinding, autonomyLevel string) error
	// GetInvestigationByFinding returns the latest investigation for a finding
	GetInvestigationByFinding(findingID string) *InvestigationSession
	// GetRunningCount returns the number of running investigations
	GetRunningCount() int
	// GetFixedCount returns the number of issues auto-fixed by Patrol
	GetFixedCount() int
	// CanStartInvestigation returns true if a new investigation can be started
	CanStartInvestigation() bool
	// ReinvestigateFinding triggers a re-investigation of a finding
	ReinvestigateFinding(ctx context.Context, findingID, autonomyLevel string) error
	// Shutdown signals all running investigations to stop, persists state,
	// and waits for them to finish (up to the context deadline).
	Shutdown(ctx context.Context) error
}

// InvestigationStoreMaintainer is an optional interface for orchestrators that
// expose their investigation store for periodic maintenance.
type InvestigationStoreMaintainer interface {
	CleanupInvestigationStore(maxAge time.Duration, maxSessions int)
}

// InvestigationFinding is the shared finding type used by the investigation
// orchestrator. Type alias so *InvestigationFinding and *finding.Finding
// are interchangeable, eliminating field-by-field copies at the adapter layer.
type InvestigationFinding = finding.Finding

// InvestigationSession represents the result of an investigation (minimal interface)
type InvestigationSession struct {
	ID             string            `json:"id"`
	FindingID      string            `json:"finding_id"`
	SessionID      string            `json:"session_id"`
	Status         string            `json:"status"`
	StartedAt      time.Time         `json:"started_at"`
	CompletedAt    *time.Time        `json:"completed_at,omitempty"`
	TurnCount      int               `json:"turn_count"`
	Outcome        string            `json:"outcome,omitempty"`
	ToolsAvailable []string          `json:"tools_available,omitempty"`
	ToolsUsed      []string          `json:"tools_used,omitempty"`
	EvidenceIDs    []string          `json:"evidence_ids,omitempty"`
	Summary        string            `json:"summary,omitempty"`
	Error          string            `json:"error,omitempty"`
	ProposedFix    *InvestigationFix `json:"proposed_fix,omitempty"`
	ApprovalID     string            `json:"approval_id,omitempty"`
}

// InvestigationFix represents a proposed remediation action
type InvestigationFix struct {
	ID          string   `json:"id"`
	Description string   `json:"description"`
	Commands    []string `json:"commands,omitempty"`
	RiskLevel   string   `json:"risk_level,omitempty"`
	Destructive bool     `json:"destructive"`
	TargetHost  string   `json:"target_host,omitempty"`
	Rationale   string   `json:"rationale,omitempty"`
}

// PatrolService runs background AI analysis of infrastructure
type PatrolService struct {
	mu sync.RWMutex

	aiService           *Service
	stateProvider       StateProvider
	thresholdProvider   ThresholdProvider
	config              PatrolConfig
	findings            *FindingsStore
	knowledgeStore      *knowledge.Store        // For per-resource notes in patrol context
	discoveryStore      *servicediscovery.Store // For AI-discovered infrastructure context
	guestProber         GuestProber             // For pre-patrol guest reachability checks
	metricsHistory      MetricsHistoryProvider  // For trend analysis and predictions
	baselineStore       *baseline.Store         // For anomaly detection via learned baselines
	changeDetector      *ChangeDetector         // For tracking infrastructure changes
	remediationLog      *RemediationLog         // For tracking remediation actions
	patternDetector     *PatternDetector        // For failure prediction from historical patterns
	correlationDetector *CorrelationDetector    // For multi-resource correlation
	incidentStore       *memory.IncidentStore   // For incident timeline capture
	alertResolver       AlertResolver           // For AI-based alert resolution

	// Unified resource provider — reads physical disks, Ceph, etc. from canonical model
	unifiedResourceProvider UnifiedResourceProvider
	// ReadState provides typed read-only views over resource state (VMs, nodes, hosts, etc.).
	// This is injected separately from stateProvider since stateProvider also contains
	// non-resource telemetry (alerts, backups, connection health) that isn't modeled as resources yet.
	readState unifiedresources.ReadState

	// New AI intelligence providers (Phase 6)
	learningProvider     LearningProvider     // For learned preferences from user feedback
	proxmoxEventProvider ProxmoxEventProvider // For recent Proxmox operations
	forecastProvider     ForecastProvider     // For trend forecasts

	// Event-driven patrol triggers (Phase 7)
	triggerManager *TriggerManager // For event-driven patrol scheduling

	// Unified intelligence facade - aggregates all subsystems for unified view
	intelligence *Intelligence

	// Circuit breaker for resilient AI API calls
	circuitBreaker *circuit.Breaker

	// Remediation engine for generating fix plans from findings
	remediationEngine *remediation.Engine

	// Investigation orchestrator for autonomous investigation of findings
	investigationOrchestrator InvestigationOrchestrator
	investigationWg           sync.WaitGroup // Tracks in-flight investigation goroutines

	// Unified findings callback - pushes findings to unified store
	unifiedFindingCallback UnifiedFindingCallback
	// Unified resolver callback - marks findings resolved in unified store
	unifiedFindingResolver func(findingID string)

	// Push notification callback - sends push via relay to mobile devices
	pushNotifyCallback PushNotifyCallback

	// Cached thresholds (recalculated when thresholdProvider changes)
	thresholds    PatrolThresholds
	proactiveMode bool // When true, warn before thresholds; when false, use exact thresholds

	// Runtime state
	running           bool
	runInProgress     bool
	runStartedAt      time.Time
	stopCh            chan struct{}
	configChanged     chan struct{} // Signal when config changes to reset ticker
	lastPatrol        time.Time
	lastDuration      time.Duration
	resourcesChecked  int
	errorCount        int
	lastBlockedReason string
	lastBlockedAt     time.Time
	nextScheduledAt   time.Time // Tracks actual next patrol time (accounts for ticker resets)

	// Patrol run history with persistence support
	runHistoryStore *PatrolRunHistoryStore

	// Ad-hoc trigger channel for event-driven patrols (alert driven)
	adHocTrigger chan *alerts.Alert

	// Live streaming support
	streamMu          sync.RWMutex
	streamSubscribers map[chan PatrolStreamEvent]*streamSubscriber
	currentOutput     streamOutputBuffer // Tail buffer for current streaming output
	streamPhase       string             // "idle", "analyzing", "complete"
	streamRunID       string             // Identifies the current streamed run (best-effort)
	streamSeq         int64              // Monotonic sequence for SSE events within streamRunID
	streamCurrentTool string             // Last observed tool name (best-effort)
	streamEvents      []PatrolStreamEvent
}

const patrolStreamMaxOutputBytes = 64 * 1024

// streamOutputBuffer retains only the most recent bytes written, to cap memory usage
// while still allowing late joiners to get a useful snapshot.
type streamOutputBuffer struct {
	buf       []byte
	truncated bool
}

func (b *streamOutputBuffer) Reset() {
	// Keep capacity bounded so long-running streams can't retain a large backing array.
	if cap(b.buf) > patrolStreamMaxOutputBytes {
		b.buf = make([]byte, 0, patrolStreamMaxOutputBytes)
	} else {
		b.buf = b.buf[:0]
	}
	b.truncated = false
}

func (b *streamOutputBuffer) Len() int             { return len(b.buf) }
func (b *streamOutputBuffer) String() string       { return string(b.buf) }
func (b *streamOutputBuffer) Truncated() bool      { return b.truncated }
func (b *streamOutputBuffer) WriteString(s string) { b.appendString(s) }

func (b *streamOutputBuffer) appendString(s string) {
	if len(s) == 0 {
		return
	}

	max := patrolStreamMaxOutputBytes
	if len(s) >= max {
		// Keep only the tail of the incoming chunk.
		b.buf = append(b.buf[:0], s[len(s)-max:]...)
		b.truncated = true
		b.normalizeUTF8Start()
		b.shrinkCapIfNeeded()
		return
	}

	// Keep as much of existing tail as possible.
	needKeep := max - len(s)
	if len(b.buf) > needKeep {
		b.buf = append(b.buf[:0], b.buf[len(b.buf)-needKeep:]...)
		b.truncated = true
	}
	b.buf = append(b.buf, s...)
	if len(b.buf) > max {
		b.buf = b.buf[len(b.buf)-max:]
		b.truncated = true
	}
	b.normalizeUTF8Start()
	b.shrinkCapIfNeeded()
}

func (b *streamOutputBuffer) normalizeUTF8Start() {
	// If we truncated by bytes, we may have cut in the middle of a UTF-8 rune.
	// Drop leading continuation bytes so the string starts on a rune boundary.
	for len(b.buf) > 0 && (b.buf[0]&0xC0) == 0x80 {
		b.buf = b.buf[1:]
		b.truncated = true
	}
}

func (b *streamOutputBuffer) shrinkCapIfNeeded() {
	if cap(b.buf) <= patrolStreamMaxOutputBytes {
		return
	}
	tmp := make([]byte, len(b.buf), patrolStreamMaxOutputBytes)
	copy(tmp, b.buf)
	b.buf = tmp
}

// ToolCallRecord captures a single tool invocation during a patrol run.
type ToolCallRecord struct {
	ID        string `json:"id"`
	ToolName  string `json:"tool_name"`
	Input     string `json:"input"`
	Output    string `json:"output"`
	Success   bool   `json:"success"`
	StartTime int64  `json:"start_time"`
	EndTime   int64  `json:"end_time"`
	Duration  int64  `json:"duration_ms"`
}

// streamSubscriber wraps a stream channel with an atomic close flag
// to prevent double-close panics when both broadcast and unsubscribe race.
type streamSubscriber struct {
	ch     chan PatrolStreamEvent
	closed atomic.Bool
	// Consecutive times we couldn't deliver to this subscriber because its channel
	// was full. Used to tolerate short bursts without immediately disconnecting.
	fullCount int
}

// PatrolStreamEvent represents a streaming update from the patrol
type PatrolStreamEvent struct {
	// Meta
	// Seq is suitable to be used as an SSE "id:" for Last-Event-ID, but replay is best-effort.
	RunID string `json:"run_id,omitempty"`
	Seq   int64  `json:"seq,omitempty"`
	TsMs  int64  `json:"ts_ms,omitempty"`
	// If this is a synthetic snapshot/resync event, why it was emitted.
	// Examples: "late_joiner", "stale_last_event_id".
	ResyncReason string `json:"resync_reason,omitempty"`
	BufferStart  int64  `json:"buffer_start_seq,omitempty"`
	BufferEnd    int64  `json:"buffer_end_seq,omitempty"`
	// True when the snapshot content has been truncated due to the tail buffer.
	ContentTruncated *bool `json:"content_truncated,omitempty"`

	// Payload
	// Known types include: "snapshot", "start", "content", "phase", "thinking",
	// "complete", "error", "tool_start", "tool_end".
	Type    string `json:"type"`
	Content string `json:"content,omitempty"`
	Phase   string `json:"phase,omitempty"`  // Current phase description
	Tokens  int    `json:"tokens,omitempty"` // Token count so far
	// Tool event fields (present only for tool_start/tool_end)
	ToolID       string `json:"tool_id,omitempty"`
	ToolName     string `json:"tool_name,omitempty"`
	ToolInput    string `json:"tool_input,omitempty"`
	ToolRawInput string `json:"tool_raw_input,omitempty"`
	ToolOutput   string `json:"tool_output,omitempty"`
	ToolSuccess  *bool  `json:"tool_success,omitempty"` // pointer so omitempty works with false
}

// NewPatrolService creates a new patrol service
func NewPatrolService(aiService *Service, stateProvider StateProvider) *PatrolService {
	return &PatrolService{
		aiService:         aiService,
		stateProvider:     stateProvider,
		config:            DefaultPatrolConfig(),
		findings:          NewFindingsStore(),
		thresholds:        DefaultPatrolThresholds(),
		stopCh:            make(chan struct{}),
		runHistoryStore:   NewPatrolRunHistoryStore(MaxPatrolRunHistory),
		streamSubscribers: make(map[chan PatrolStreamEvent]*streamSubscriber),
		streamPhase:       "idle",
		adHocTrigger:      make(chan *alerts.Alert, 10), // Buffer triggers
	}
}
